bgpd/bgp_mplsvpn.c

```diff
@@ -1562,34 +1563,63 @@ void vpn_leak_from_vrf_update(struct bgp *to_bgp,	     /* to */
 
 	/* Set SID for SRv6 VPN */
 	if (from_bgp->vpn_policy[afi].tovpn_sid_locator) {
+		struct listnode *node;
+		struct srv6_locator_chunk *chunk;
+		for (ALL_LIST_ELEMENTS_RO(to_bgp->srv6_locator_chunks, node,
+					  chunk))
+			if (!strcmp(to_bgp->srv6_locator_name,
+				    chunk->locator_name))
+				break;
+
 		encode_label(
 			from_bgp->vpn_policy[afi].tovpn_sid_transpose_label,
 			&label);
+
 		static_attr.srv6_l3vpn = XCALLOC(MTYPE_BGP_SRV6_L3VPN,
 				sizeof(struct bgp_attr_srv6_l3vpn));
 		static_attr.srv6_l3vpn->sid_flags = 0x00;
-		static_attr.srv6_l3vpn->endpoint_behavior = 0xffff;
-		static_attr.srv6_l3vpn->loc_block_len =
-			from_bgp->vpn_policy[afi]
-				.tovpn_sid_locator->block_bits_length;
-		static_attr.srv6_l3vpn->loc_node_len =
-			from_bgp->vpn_policy[afi]
-				.tovpn_sid_locator->node_bits_length;
+		static_attr.srv6_l3vpn->endpoint_behavior =
+			afi == AFI_IP
+				? IANA_SEG6_ENDPOINT_BEHAVIOR_END_DT4
+				: (afi == AFI_IP6
+					   ? IANA_SEG6_ENDPOINT_BEHAVIOR_END_DT6
+					   : 0xffff);
+		if (chunk) {
+			static_attr.srv6_l3vpn->loc_block_len =
+				chunk->block_bits_length;
+			static_attr.srv6_l3vpn->loc_node_len =
+				chunk->node_bits_length;
+			static_attr.srv6_l3vpn->transposition_len =
+				chunk->function_bits_length ?: 16;
+			static_attr.srv6_l3vpn->transposition_offset =
+				chunk->block_bits_length +
+				chunk->node_bits_length;
+		} else {
+			static_attr.srv6_l3vpn->loc_block_len =
+				from_bgp->vpn_policy[afi]
+					.tovpn_sid_locator->block_bits_length;
+			static_attr.srv6_l3vpn->loc_node_len =
+				from_bgp->vpn_policy[afi]
+					.tovpn_sid_locator->node_bits_length;
+			static_attr.srv6_l3vpn->transposition_len =
+				from_bgp->vpn_policy[afi]
+					.tovpn_sid_locator
+					->function_bits_length;
+			static_attr.srv6_l3vpn->transposition_offset =
+				from_bgp->vpn_policy[afi]
+					.tovpn_sid_locator->block_bits_length +
+				from_bgp->vpn_policy[afi]
+					.tovpn_sid_locator->node_bits_length;
+			;
+		}
```

Comment:
Why do you re-calculate these parameters?

The SRv6 locator from which SRv6 SID is allocated is saved in `from_bgp->vpn_policy[afi].tovpn_sid_locator`. So, you don't have to find locators and use them instead of saved ones.

And, current SRv6 locators have function length because user-specified or default function length is saved in srv6_locator structure. We don't have to consider the default function length.

### References

- https://github.com/FRRouting/frr/blob/master/bgpd/bgp_mplsvpn.c#L770-L791
- https://github.com/FRRouting/frr/blob/master/bgpd/bgp_mplsvpn.c#L679-L680
- https://github.com/FRRouting/frr/blob/master/zebra/zebra_srv6_vty.c#L274-L370

Reply:
Thank you for your feedback. You are correct. I just submitted the update.